### PR TITLE
Display event month view branch

### LINKED ISSUE #29: FE - Set up Day View for Calendar page
### Requirements:
- [ ] Take a look to this github (https://github.com/chaseottofy/google-calendar-clone-vanilla) as a reference
- [ ] implement getDay function to create an array of dayjs objects, representing time slots by hour with 15-minute intervals (from 00:00 to 23:45)
- [ ] display day view 
- [ ] check what time it is so there will be a horizontal line marking the current time (similar to isToday)

### Dependencies:
- frontend/components/Calendar
- frontend/service
- frontend/pages

### Done requirements:
- Functioning as expected

### LINKED ISSUE #31: FE - Display Events on Month View of Calendar
### Requirements:
- [x] create function to store events for a day
- [x] render those events on the month view (loop through all events and days)
- [x] style event display
- [x] implement event handlers and other functionality?

### Dependencies:
- frontend/components/Calendar

### Done requirements:
- functioning as expected

### LINKED ISSUE #37: FE + BE - Display Event on Month View
### Requirements:
- [x] please list what tasks need to be done (be specific here)
- [x] Create new branch based on "newDayView_invitedGuest_PhuocDang" branch
- [x] Create getAllEventToday() and algorithm to fetch all today events in database in Event.js controller
- [x] Create getAllEventMonths() and algorithm to fetch all Month events in database in Event.js controller
- [x] Set up Redux Toolkit to make the data always up-to-date (so MonthEvents is global state, and "set/add/delete/update" reducer to keep this state always up-to-date)
- [x] Mapping MonthEvents into <div>{event.title}</div> to show it in components/Calendar/Day.js
- [x] Work Flow to display events on Month Views:

1. Rendering Month View -> Send GET Request to backend to trigger getAllEventMonths() and assign those data into MonthEvents in Redux Store -> In Day.js component, each of day cell filter the events from MonthEvents to get specific events taking place on that day -> Then show all of them on screen. 

2. When we move to next/previous months -> it will re-render MonthView -> Then re-trigger getAllEventMonths() to update MonthEvents again

3. When we create/update/add/remove events by <EventForm/>, it will trigger set/add/delete/update reducers in Redux Store and adjust MonthEvents smoothly on screen without needing to re-fetch the database to update MonthEvents again

### Dependencies:
- Please list the files that might be involved in this issue
- components/Calendar/Day.js
- backend/controllers/Event.js
- store/store.js
- frontend/slice
### Done requirements:
- When will this issue count as done?
- When it can display Events on MonthView, and automatically update whenever we createEvent()/deleteEvent()/modifyEvent()

### PR DESCRIPTION
**Pull request by PhuocOng**
Create new branch based on "newDayView_invitedGuest_PhuocDang" branch
 Create getAllEventToday() and algorithm to fetch all today events in database in Event.js controller
 Create getAllEventMonths() and algorithm to fetch all Month events in database in Event.js controller
 Set up Redux Toolkit to make the data always up-to-date (so MonthEvents is global state, and "set/add/delete/update" reducer to keep this state always up-to-date)
 Mapping MonthEvents into {event.title} to show it in components/Calendar/Day.js
 Work Flow to display events on Month Views:
1. Rendering Month View -> Send GET Request to backend to trigger getAllEventMonths() and assign those data into MonthEvents in Redux Store -> In Day.js component, each of day cell filter the events from MonthEvents to get specific events taking place on that day -> Then show all of them on screen.

2. When we move to next/previous months -> it will re-render MonthView -> Then re-trigger getAllEventMonths() to update MonthEvents again

3. When we create/update/add/remove events by , it will trigger set/add/delete/update reducers in Redux Store and adjust MonthEvents smoothly on screen without needing to re-fetch the database to update MonthEvents again

https://github.com/viettech-group3/productify/assets/122703392/80f6910e-767e-4886-be84-0c5fd2e13162

### CODE REVIEW

File: backend/controllers/Event.js [hunk 1]
```diff
@@ -36,7 +37,7 @@ const createEvent = async (req, res) => {
       //Make sure all promise is fullfilled
       await Promise.all(invitedParticipation);
     }
-    res.status(201).json({ event: event });
+    res.status(201).json(event.toJSON());
```
Comment (khoa165) [inline backend/controllers/Event.js]: are you sure this is needed?
`json({ event: event })` will look like
```
data: {
  event: {
    name: ...,
    description: ...,
    etc
  }
}
```
And I might be wrong but `json(event)` will be enough for this
```
data: {
  name: ...,
  description: ...,
  etc
}
```

File: backend/controllers/Event.js [hunk 2]
```diff
@@ -83,6 +83,92 @@ const getAllEvents = async (req, res) => {
...
+    const todayEvents = eventsOfThisUser.filter(event => {
+      const start = new Date(event.start);
+      const end = new Date(event.end);
+      const startOfCurrentDate = new Date(currentDate.getFullYear(), currentDate.getMonth(), currentDate.getDate(), 0, 0, 0, 1);
+      const endOfCurrentDate = new Date(currentDate.getFullYear(), currentDate.getMonth(), currentDate.getDate(), 23, 59, 59, 999);
+      console.log(start, end, startOfCurrentDate, endOfCurrentDate, "Date is: ", event.start.toString(), event.end.toString())
+      const minTime = Math.min(start.getTime(), end.getTime(), startOfCurrentDate.getTime(), endOfCurrentDate.getTime())
+      const maxTime = Math.max(start.getTime(), end.getTime(), startOfCurrentDate.getTime(), endOfCurrentDate.getTime())
+      return ((maxTime - minTime) < 24 * 60 * 60 * 1000 + (end.getTime() - start.getTime()));
+    });
```
Comment (khoa165) [inline backend/controllers/Event.js]: can we abstract this and move to util? It makes the controller file look pretty messy. Also, it seems you try to write the logic yourself. Some library has function like `isAfter` or `isBefore` or some other functions that could you help simplify some of this

File: backend/controllers/Event.js [hunk 3]
```diff
@@ -83,6 +83,92 @@ const getAllEvents = async (req, res) => {
...
+    );
+    const flattenedEvents = eventsOfThisUser.flat(); // Flattens the array of arrays (of this user's events), make 2D array become 1D array
+    const monthEvents = flattenedEvents.filter(event => {
+      const startOfEvent = new Date(event.start);
+      const endOfEvent = new Date(event.end);
+      const startOfCurrentMonth = new Date(startDate.getFullYear(), startDate.getMonth(), startDate.getDate(), 0, 0, 0, 1);
+      const endOfCurrentMonth = new Date(endDate.getFullYear(), endDate.getMonth(), endDate.getDate(), 23, 59, 59, 999);
+      console.log(startOfEvent, endOfEvent, startOfCurrentMonth, endOfCurrentMonth, "Date is: ", event.start.toString(), event.end.toString())
+      return (startOfEvent <= endOfCurrentMonth && endOfEvent >= startOfCurrentMonth);
+    });
```
Comment (khoa165) [inline backend/controllers/Event.js]: same comment as above

File: frontend/src/components/Calendar/CalendarHeader.js [hunk 4]
```diff
@@ -1,37 +1,95 @@
...
+          height="16"
+          fill="currentColor"
+          className={' bi bi-chevron-left'}
+          viewBox="0 0 16 16"
+        >
+          <path
+            fillRule="evenodd"
+            d="M11.354 1.646a.5.5 0 0 1 0 .708L5.707 8l5.647 5.646a.5.5 0 0 1-.708.708l-6-6a.5.5 0 0 1 0-.708l6-6a.5.5 0 0 1 .708 0z"
+          />
+        </svg>
```
Comment (khoa165) [inline frontend/src/components/Calendar/CalendarHeader.js]: can we make this a separate component?

File: frontend/src/components/Calendar/CalendarHeader.js [hunk 5]
```diff
@@ -1,37 +1,95 @@
...
+          height="16"
+          fill="currentColor"
+          className={' bi bi-chevron-right'}
+          viewBox="0 0 16 16"
+        >
+          <path
+            fillRule="evenodd"
+            d="M4.646 1.646a.5.5 0 0 1 .708 0l6 6a.5.5 0 0 1 0 .708l-6 6a.5.5 0 0 1-.708-.708L10.293 8 4.646 2.354a.5.5 0 0 1 0-.708z"
+          />
+        </svg>
```
Comment (khoa165) [inline frontend/src/components/Calendar/CalendarHeader.js]: same, please extract it into its own component

File: frontend/src/components/Calendar/Day.js [hunk 6]
```diff
@@ -1,28 +1,41 @@
...
+  const MonthEvents = useSelector(state => state.MonthEvents.value)
+  const todayEvents = MonthEvents.filter(event => {  //Algorithm to filter which events is on today
+    const start = new Date(event.start);
+    const end = new Date(event.end);
+    const startOfCurrentDate = new Date(currentDate.getFullYear(), currentDate.getMonth(), currentDate.getDate(), 0, 0, 0, 1);
+    const endOfCurrentDate = new Date(currentDate.getFullYear(), currentDate.getMonth(), currentDate.getDate(), 23, 59, 59, 999);
+    const minTime = Math.min(start.getTime(), end.getTime(), startOfCurrentDate.getTime(), endOfCurrentDate.getTime())
+    const maxTime = Math.max(start.getTime(), end.getTime(), startOfCurrentDate.getTime(), endOfCurrentDate.getTime())
+    return ((maxTime - minTime) < 24 * 60 * 60 * 1000 + (end.getTime() - start.getTime()));
+  });
```
Comment (khoa165) [inline frontend/src/components/Calendar/Day.js]: It's a good practice to separate the logic from the rendering, so best to extract it to a helper/util file

File: frontend/src/components/Calendar/DayView/DayView.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+                        <div>2 PM</div>
+                        <div>3 PM</div>
+                        <div>4 PM</div>
+                        <div>5 PM</div>
+                        <div>6 PM</div>
+                        <div>7 PM</div>
+                        <div>8 PM</div>
+                        <div>9 PM</div>
+                        <div>10 PM</div>
+                        <div>11 PM</div>
```
Comment (khoa165) [inline frontend/src/components/Calendar/DayView/DayView.js]: Could we make this look cleaner with loop or something?

**Review by khoa165 — CHANGES_REQUESTED** [review]
I left some feedback on certain code style. Also make sure indentation is consistent, half of your files have 4 spaces indentation, the other half have 2 spaces indentation

File: backend/controllers/Event.js (hunk 1 above)
Comment (PhuocOng) [reply backend/controllers/Event.js]: Oh my bad, appreciate your feedback!

File: frontend/src/components/Calendar/CalendarHeader.js (hunk 4 above)
Comment (PhuocOng) [reply frontend/src/components/Calendar/CalendarHeader.js]: Of course! I just put all small icons into one icons.js file in frontend/src/assets

File: backend/controllers/Event.js (hunk 2 above)
Comment (PhuocOng) [reply backend/controllers/Event.js]: Roger that, I will try to shorten them

File: frontend/src/components/Calendar/Day.js (hunk 6 above)
Comment (PhuocOng) [reply frontend/src/components/Calendar/Day.js]: Got it, I've done with them

**Review by lhoang25 — APPROVED** [review]

**Pull request merged**